rust/ql/test/library-tests/type-inference/main.rs

```diff
@@ -2459,6 +2459,732 @@ pub mod pattern_matching {
+                let arr_b = b; // $ MISSING: type=arr_b:i32
+                let arr_c = c; // $ MISSING: type=arr_c:i32
+                println!("Array elements: {}, {}, {}", arr_a, arr_b, arr_c);
+            }
+        }
+    }
+
+    pub fn path_patterns() {
+        // PathPat - Path patterns (for enums, constants, etc.)
+        const CONSTANT: i32 = 42;
+        let value = 42i32;
+
+        match value {
+            CONSTANT => {
+                let const_match = value; // $ type=const_match:i32
+                println!("Matches constant: {}", const_match);
+            }
+            _ => {}
+        }
+
+        // Enum variants as path patterns
+        let option = MyOption::Some(10i32);
+        match option {
+            MyOption::None => {
+                println!("None variant");
+            }
+            MyOption::Some(x) => {
+                let some_value = x; // $ type=some_value:i32
+                println!("Some value: {}", some_value);
+            }
+        }
+
+        // Module path patterns
+        match std::result::Result::Ok::<i32, usize>(42) {
+            std::result::Result::Ok(x) => {
+                let ok_value = x; // $ type=ok_value:i32
+                println!("Ok value: {}", ok_value);
+            }
+            std::result::Result::Err(e) => {
+                let err_value = e; // $ type=err_value:usize
+                println!("Error: {}", err_value);
+            }
+        }
+    }
+
+    pub fn or_patterns() {
+        let value = 42i32;
+
+        // OrPat - Or patterns
+        match value {
+            1 | 2 | 3 => {
+                let small_num = value; // $ type=small_num:i32
+                println!("Small number: {}", small_num);
+            }
+            10 | 20 => {
+                let round_num = value; // $ type=round_num:i32
+                println!("Round number: {}", round_num);
+            }
+            _ => {}
+        }
+
+        // Complex or pattern with structs
+        let point = Point { x: 0, y: 5 };
+        match point {
+            Point { x: x @ 0, y } | Point { x, y: y @ 0 } => {
+                let axis_x = x; // $ type=axis_x:i32
+                let axis_y = y; // $ type=axis_y:i32
+                println!("Point on axis: ({}, {})", axis_x, axis_y);
+            }
+            _ => {}
+        }
+
+        // Or pattern with ranges
+        match value {
+            1..=10 | 90..=100 => {
+                let range_or_value = value; // $ type=range_or_value:i32
+                println!("In range: {}", range_or_value);
+            }
+            _ => {}
+        }
+    }
+
+    pub fn box_patterns() {
+        let boxed_value = Box::new(100i32); // $ method=new
+
+        // BoxPat - Box patterns (requires feature flag)
+        match boxed_value {
+            box 100 => {
+                println!("Boxed 100");
+            }
+            box x => {
+                let unboxed = x; // $ MISSING: type=unboxed:i32
```

Comment:
@hvitved Box patterns are probably not very often used, but it might be easy enough to just add support for them.